### PR TITLE
Always emit Java 8 bytecode on 2.12, regardless of `-release`; deprecate `-target`

### PR DESCRIPTION
**Pull request by lrytz**
Unlike Scala 2.13, Scala 2.12 cannot emit valid class files for target bytecode versions newer than 8. Thus, this PR deprecates `-target` on 2.12.

Regardless, you may still use `-release` to compile against a specific platform API version.

For more information on `-release`, see the Scala 2.13 PR #9982. This PR, taken together with #10089, effectively backports 9982 to 2.12, with the exception that on 2.12, bytecode version 8 is always emitted.

### CODE REVIEW

**Comment by som-snytt** [comment]
Why isn't `target` tied to `8`, with an explanatory message if they try to change it?

I almost wrote 'exoplanetary". That is for messages from aliens.

**Comment by lrytz** [comment]
👍  good suggestion, I added that

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala [hunk 1]
```diff
@@ -71,15 +71,17 @@ trait StandardScalaSettings { _: MutableSettings =>
...
-        if (SupportedTargetVersions.contains(setting.value))
-          setting.withDeprecationMessage("Use -release instead to compile against the correct platform API.")
-        else {
+        if (setting.value.toInt > MaxSupportedTargetVersion) {
+          setting.withDeprecationMessage(s"Scala 2.12 cannot emit valid class files for targets newer than $MaxSupportedTargetVersion (this is possible with Scala 2.13). Use -release to compile against a specific platform API version.")
+          setting.value = DefaultTargetVersion
+        } else if (setting.value.toInt < MinSupportedTargetVersion) {
           setting.withDeprecationMessage(s"${setting.name}:${setting.value} is deprecated, forcing use of $DefaultTargetVersion")
-          setting.value = DefaultTargetVersion   // triggers this hook
+          setting.value = DefaultTargetVersion
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:79]: Is the comment falsified? I see L73 as why it doesn't loop.

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala [hunk 2]
```diff
@@ -71,15 +71,17 @@ trait StandardScalaSettings { _: MutableSettings =>
...
         }
     }
     .withAbbreviation("--target")
-  def targetValue: String = releaseValue.getOrElse(target.value)
+  // Unlike 2.13, don't use `releaseValue.getOrElse(target.value)`, because 2.12 doesn't have a fix for scala-dev#408
+  def targetValue: String = target.value
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:84]: If they use just `-release` then they get default target 8. If they `-release 8 -target 6` they will get the existing deprecation and target 8. If they are building on JDK 8, they can only use `-release 8` or use `-target`. The changed behavior is `-release 11` where they get target 8 as a limitation with no warning to fail a build. Only explicit `-target 9` would warn.

**Review by som-snytt — APPROVED** [review]
That `.withPreSetHook(normalizeTarget)` really shines when you just want to `value.toInt`.

File: src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala (hunk 1 above)
Comment (lrytz) [reply src/compiler/scala/tools/nsc/settings/StandardScalaSettings.scala:79]: Ah, no, I think I just found it distracting...

**Pull request merged**